quickwit-search/src/fetch_docs.rs

```diff
@@ -144,30 +163,140 @@ async fn get_searcher_for_split_without_cache(
     Ok(reader)
 }
 
+/// A struct for holding a fetched document's content and highlights.
+struct DocBundle {
+    doc_json: String,
+    highlights_json: Option<String>,
+}
+
 /// Fetching docs from a specific split.
 #[tracing::instrument(skip(global_doc_addrs, index_storage, split))]
 #[allow(clippy::needless_lifetimes)]
 async fn fetch_docs_in_split(
     mut global_doc_addrs: Vec<GlobalDocAddress>,
     index_storage: Arc<dyn Storage>,
     split: &SplitIdAndFooterOffsets,
-) -> anyhow::Result<Vec<(GlobalDocAddress, String)>> {
+    doc_mapper: Arc<dyn DocMapper>,
+    search_request: &SearchRequest,
+) -> anyhow::Result<Vec<(GlobalDocAddress, DocBundle)>> {
     global_doc_addrs.sort_by_key(|doc| doc.doc_addr);
 
     let index_reader = get_searcher_for_split_without_cache(index_storage, split).await?;
-    let searcher = Arc::new(index_reader.searcher());
+    let searcher = index_reader.searcher();
+    let snippet_generators =
+        Arc::new(create_snippet_generators(&searcher, doc_mapper, search_request).await?);
+
     let doc_futures = global_doc_addrs.into_iter().map(|global_doc_addr| {
         let searcher = searcher.clone();
+        let moved_snippet_generators = snippet_generators.clone();
         async move {
             let doc = searcher
                 .doc_async(global_doc_addr.doc_addr)
                 .await
                 .context("searcher-doc-async")?;
             let doc_json = searcher.schema().to_json(&doc);
-            Ok((global_doc_addr, doc_json))
+            if moved_snippet_generators.is_empty() {
+                return Ok((
+                    global_doc_addr,
+                    DocBundle {
+                        doc_json,
+                        highlights_json: None,
+                    },
+                ));
+            }
+
+            let mut highlight = HashMap::new();
+            for (field, field_values) in doc.get_sorted_field_values() {
+                let field_name = searcher.schema().get_field_name(field);
+                if let Some(snippet_generator) = moved_snippet_generators.get(field_name) {
+                    let values = snippets_from_field_values(snippet_generator, field_values);
```

Comment:
I wonder if we can avoid collecting to a vector here since we're serializing to JSON right after this step.

Reply:
I thought of this, but it involves implementing the serialization for the iterator and this seems to be discouraged by serde since serializing an iterator twice could yield different results.